caluma/caluma_form/structure.py

```diff
@@ -0,0 +1,212 @@
+"""Hierarchical representation of a document / form."""
+import weakref
+from functools import singledispatch
+from typing import Optional
+
+from .models import Question
+
+
+def object_local_memoise(meth):
+    def new_meth(self, *args, **kwargs):
+        if not hasattr(self, "_memoise"):
+            self._memoise = {}
+
+        key = str([args, kwargs, meth])
+        if key in self._memoise:
+            return self._memoise[key]
+        ret = meth(self, *args, **kwargs)
+        self._memoise[key] = ret
+        return ret
+
+    return new_meth
+
+
+class Element:
+    def __init__(self, parent=None):
+        self._parent = weakref.ref(parent) if parent else None
+
+    def parent(self):
+        return self._parent() if self._parent else None
+
+    def children(self):  # pragma: no cover
+        return []
+
+
+class AnsQuestion(Element):
```

Comment:
In `ember-caluma` we're using the same idea, but we're calling it `field` - might be a better name than `AnsQuestion`?

Reply:
Updated the naming